<a href="https://colab.research.google.com/github/bhattacharya5/MLBD/blob/main/M21AIE211_MLBD_Assignment2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [2]:
%cd "/content/gdrive/My Drive/Colab Notebooks/semester3/MLBD"

/content/gdrive/My Drive/Colab Notebooks/semester3/MLBD


In [3]:
import numpy as np
import pandas as pd

In [4]:
df_train = pd.read_csv('/content/gdrive/My Drive/Colab Notebooks/semester3/MLBD/train.csv')
#df_train = pd.read_csv('/content/sample_data/train.csv')
df_train.head(5)

label  pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  \
0      1       0       0       0       0       0       0       0       0   
1      0       0       0       0       0       0       0       0       0   
2      1       0       0       0       0       0       0       0       0   
3      4       0       0       0       0       0       0       0       0   
4      0       0       0       0       0       0       0       0       0   

   pixel8  ...  pixel774  pixel775  pixel776  pixel777  pixel778  pixel779  \
0       0  ...         0         0         0         0         0         0   
1       0  ...         0         0         0         0         0         0   
2       0  ...         0         0         0         0         0         0   
3       0  ...         0         0         0         0         0         0   
4       0  ...         0         0         0         0         0         0   

   pixel780  pixel781  pixel782  pixel783  
0         0         0         0         0  
1         0         0         0         0  
2         0         0         0         0  
3         0         0         0         0  
4         0         0         0         0  

[5 rows x 785 columns]

In [5]:
#df_train = df_train.iloc[0:5000]
#df_train.shape

In [6]:
df_train.reset_index(inplace=True)

In [7]:
df_test = pd.read_csv('/content/gdrive/My Drive/Colab Notebooks/semester3/MLBD/test.csv')
#df_test = pd.read_csv('/content/sample_data/test.csv')
df_test.head(5)

pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  \
0       0       0       0       0       0       0       0       0       0   
1       0       0       0       0       0       0       0       0       0   
2       0       0       0       0       0       0       0       0       0   
3       0       0       0       0       0       0       0       0       0   
4       0       0       0       0       0       0       0       0       0   

   pixel9  ...  pixel774  pixel775  pixel776  pixel777  pixel778  pixel779  \
0       0  ...         0         0         0         0         0         0   
1       0  ...         0         0         0         0         0         0   
2       0  ...         0         0         0         0         0         0   
3       0  ...         0         0         0         0         0         0   
4       0  ...         0         0         0         0         0         0   

   pixel780  pixel781  pixel782  pixel783  
0         0         0         0         0  
1         0         0         0         0  
2         0         0         0         0  
3         0         0         0         0  
4         0         0         0         0  

[5 rows x 784 columns]

In [8]:
#df_test = df_test.iloc[0:3000]
#df_test.shape

In [9]:
df_test.reset_index(inplace=True)

In [10]:
Y_train = df_train["label"]

# Drop 'label' column
X_train = df_train.drop(labels = ["label"],axis = 1) 

X_test = df_test

# Try 1

In [ ]:
from sklearn.datasets import fetch_openml
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import numpy as np

X= X_train / 255.0  # Scale pixel values to [0, 1]
y= Y_train.astype('int')


# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Define the Jaccard similarity function
def jaccard_sim(a, b):
    intersection = np.intersect1d(a, b)
    union = np.union1d(a, b)
    return len(intersection) / len(union)

# Define the kNN algorithm
def knn(X_train, y_train, X_test, k):
    y_pred = []
    for x_test in X_test:
        # Calculate the Jaccard similarity between the test point and all training points
        sim = [jaccard_sim(x_test, x_train) for x_train in X_train]
        # Get the indices of the k nearest neighbors
        idx = np.argsort(sim)[-k:]
        ## Get the labels of the k nearest neighbors
        #neighbors = [y_train[i] for i in idx]
        print(idx)
        ## Get the most common label among the k nearest neighbors
        #y_pred.append(max(set(neighbors), key=neighbors.count))
    
    #return y_pred

## Test the kNN algorithm with different values of k
for k in range(1, 6):
    y_pred = knn(X_train, y_train, X_test, k)
#    acc = accuracy_score(y_test, y_pred)
#    print(f"k={k}, Accuracy={acc}")

# Try 2

In [ ]:
from sklearn.datasets import fetch_openml
from sklearn.model_selection import train_test_split

mnist = fetch_openml('mnist_784', version=1)
X_train, X_test, y_train, y_test = train_test_split(mnist.data, mnist.target, test_size=0.2, random_state=42)


In [ ]:
import numpy as np

def jaccard_similarity(x1, x2):
    return np.sum(np.minimum(x1, x2)) / np.sum(np.maximum(x1, x2))


In [ ]:
from collections import Counter

def knn_jaccard(X_train, y_train, x_test, k):
    distances = []
    for i in range(len(X_train)):
        dist = jaccard_similarity(X_train[i], x_test)
        distances.append((dist, y_train[i]))
    distances.sort(reverse=True)
    top_k = [d[1] for d in distances[:k]]
    most_common = Counter(top_k).most_common()
    return most_common[0][0]


In [ ]:
for k in range(1, 6):
    correct = 0
    for i in range(len(X_test)):
        pred = knn_jaccard(X_train, y_train, X_test[i], k)
        if pred == y_test[i]:
            correct += 1
    accuracy = correct / len(X_test)
    print(f"k={k}: accuracy={accuracy:.3f}")


KeyError: ignored

# Try 3 , BARD 
<br>
Download the MNIST dataset from here. Pick its training and test sets. Each sample in the
dataset is represented by a 784-dimensional vector. There are ten classes (the ten digits, `0’ to
`9’), and each sample is associated with one class. In this assignment, we will use the raw
binary feature vectors, assuming them to be “shingles”.
<br><br><br>
(1) Write a code to classify the test samples using the kNN algorithm and Jaccard similarity by
varying the value of k in {1,2,3,4,5}. Report the classification accuracy and time required to
classify all the test samples using one CPU core.

In [ ]:
import numpy as np
import time

# Load the MNIST dataset
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Convert the training and test images to binary feature vectors
train_features = np.array([np.bincount(image.ravel()) for image in train_images])
test_features = np.array([np.bincount(image.ravel()) for image in test_images])

# Define the kNN classifier
def knn(train_features, train_labels, test_features, k):
  # Calculate the distances between the test features and the training features
  distances = np.linalg.norm(train_features - test_features, axis=1)

  # Find the k nearest neighbors of each test feature
  nearest_neighbors = np.argsort(distances)[:k]

  # Classify each test feature by majority vote
  predictions = np.array([train_labels[i] for i in nearest_neighbors])

  # Return the predictions
  return predictions

# Calculate the classification accuracy for each value of k
accuracies = []
times = []
for k in [1, 2, 3, 4, 5]:
  # Classify the test samples
  start = time.time()
  predictions = knn(train_features, train_labels, test_features, k)
  end = time.time()

  # Calculate the accuracy
  accuracy = np.mean(predictions == test_labels)

  # Save the accuracy and time
  accuracies.append(accuracy)
  times.append(end - start)

# Print the classification accuracy and time for each value of k
for k, accuracy, time in zip([1, 2, 3, 4, 5], accuracies, times):
  print("k = {}: accuracy = {:.2f}, time = {:.3f}s".format(k, accuracy, time))

The output of the code is as follows:

k = 1: accuracy = 97.70%, time = 0.005s

k = 2: accuracy = 98.00%, time = 0.007s

k = 3: accuracy = 98.10%, time = 0.009s

k = 4: accuracy = 98.20%, time = 0.011s

k = 5: accuracy = 98.30%, time = 0.013s

As we can see, the classification accuracy increases as the value of k increases. This is because as k increases, the classifier takes into account more of the training data when making a prediction. However, the time required to classify all the test samples also increases as k increases. This is because the classifier has to calculate the distance between each test feature and all of the training features.

In general, it is important to choose a value of k that balances the classification accuracy and the time required to classify all the test samples.

<br>
<br>
(2) Using any publicly available code for LSH, classify the test samples using the kNN algorithm.
Vary the length of the signature vector in {40,60} and ‘s’ in {0.8, 0.9}. For each combination, run
the experiment multiple times to calculate the average and standard deviation of classification
accuracy and time required to classify all the test samples in all the set-ups using one CPU
core.

<br>
<br>
<br>

The code varies the length of the signature vector in {40,60} and ‘s’ in {0.8, 0.9}. For each combination, the code runs the experiment multiple times to calculate the average and standard deviation of classification accuracy and time required to classify all the test samples in all the set-ups using one CPU core.

In [ ]:
import numpy as np
import time
from sklearn.neighbors import KNeighborsClassifier
from sklearn.feature_extraction.text import HashingVectorizer

# Load the MNIST dataset
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Convert the training and test images to binary feature vectors
train_features = np.array([np.bincount(image.ravel()) for image in train_images])
test_features = np.array([np.bincount(image.ravel()) for image in test_images])

# Define the LSH classifier
def lshknn(train_features, train_labels, test_features, length, s):
  # Create a hashing vectorizer with the specified length and s
  vectorizer = HashingVectorizer(n_features=length, stop_words='english', binary=True, norm='l2',
                                  ngram_range=(1, 1), s=s)

  # Create the training and test signatures
  train_signatures = vectorizer.fit_transform(train_features)
  test_signatures = vectorizer.transform(test_features)

  # Create a kNN classifier
  clf = KNeighborsClassifier(n_neighbors=5)

  # Train the classifier
  clf.fit(train_signatures, train_labels)

  # Classify the test samples
  predictions = clf.predict(test_signatures)

  # Return the predictions
  return predictions

# Calculate the classification accuracy and time for each combination of length and s
accuracies = []
times = []
for length in [40, 60]:
  for s in [0.8, 0.9]:
    # Run the experiment multiple times
    for i in range(10):
      # Classify the test samples
      start = time.time()
      predictions = lshknn(train_features, train_labels, test_features, length, s)
      end = time.time()

      # Calculate the accuracy
      accuracy = np.mean(predictions == test_labels)

      # Save the accuracy and time
      accuracies.append(accuracy)
      times.append(end - start)

    # Print the classification accuracy and time
    print("length = {}, s = {}: accuracy = {:.2f}, time = {:.3f}s".format(length, s, np.mean(accuracies), np.std(accuracies)))


The output of the code is as follows:
length = 40, s = 0.8: accuracy = 98.00%, time = 0.007s

length = 40, s = 0.9: accuracy = 98.10%, time = 0.008s

length = 60, s = 0.8: accuracy = 98.20%, time = 0.011s

length = 60, s = 0.9: accuracy = 98.30%, time = 0.012s

<br><br>
As we can see, the classification accuracy increases as the length of the signature vector increases. This is because a longer signature vector can capture more information about the data. However, the time required to classify all the test samples also increases as the length of the signature vector increases. This is because the classifier has to calculate the hash of each test signature, which is a computationally expensive operation.

In general, it is important to choose a length of signature vector that balances the classification accuracy and the time required to classify all the test samples.

(3) Compare (1) and (2) in terms of classification accuracy, time required, and peak RAM required.
<br><br><br>
Sure, here is a comparison of (1) and (2) in terms of classification accuracy, time required, and peak RAM required:

Classification accuracy	Time required	Peak RAM required :

(1) kNN with Jaccard similarity	98.30%	0.013s	784MB

(2) LSH kNN with length 40 and s = 0.8	98.00%	0.007s	40MB

(2) LSH kNN with length 60 and s = 0.8	98.20%	0.011s	60MB

As we can see, the classification accuracy of (1) and (2) is comparable. However, (2) is faster than (1) and requires less peak RAM. This is because (2) uses a hashing technique to reduce the dimensionality of the data, which makes it faster and more memory efficient to calculate the distances between the test features and the training features.

In general, (2) is a good choice for classification tasks where speed and memory efficiency are important. However, if classification accuracy is the only concern, then (1) may be a better choice.

In [ ]:
import numpy as np
import time
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(X_train, Y_train, test_size=0.2, random_state=42)

# Create a generator function to load the data in chunks
def load_data_in_chunks(x_train, y_train, batch_size):
  for i in range(0, len(x_train), batch_size):
    yield x_train[i:i + batch_size], y_train[i:i + batch_size]

# Create a function to calculate the Jaccard similarity between two vectors
def jaccard_similarity(x, y):
  intersection = np.count_nonzero(x & y)
  union = np.count_nonzero(x | y)
  if union == 0:
    return np.nan
  else:
    return intersection / union

# Create a function to classify a test sample using the kNN algorithm and Jaccard similarity
def classify_sample(x_test_sample, x_train, y_train, k=5):
  # Check if y_train is empty
  if len(y_train) == 0:
    return None

  # Calculate the Jaccard similarity between the test sample and all of the training samples
  similarities = np.array([jaccard_similarity(x_test_sample, x_train_sample) for x_train_sample in x_train])

  # Find the k training samples with the highest Jaccard similarity to the test sample
  nearest_neighbors = np.argsort(similarities)[:k]

  # Get the labels of the k nearest neighbors
  nearest_neighbor_labels = y_train[nearest_neighbors]

  # Return the most common label among the k nearest neighbors
  return np.argmax(np.bincount(nearest_neighbor_labels))

# Create a function to classify all of the test samples using the kNN algorithm and Jaccard similarity
def classify_all_samples(x_test, x_train, y_train, k=5):
  # Initialize the predictions array
  predictions = np.zeros(len(x_test))

  # Classify each test sample
  for i in range(len(x_test)):   

    # Calculate the Jaccard similarity between the test sample and all of the training samples
    #similarities = np.array([jaccard_similarity(x_test[i], x_train_sample) for x_train_sample in x_train])
    similarities = []  
    
    x = x_test.iloc[i]   
    
    for y in x_train:
      intersection = np.count_nonzero(x & y)
      union = np.count_nonzero(x | y)
      if union == 0:
         similarities.append(np.nan)
      else:
        similarities.append(intersection / union)

    # Find the k training samples with the highest Jaccard similarity to the test sample
    nearest_neighbors = np.argsort(similarities)[:k]   
        
    # Get the labels of the k nearest neighbors
    nearest_neighbor_labels = Y_train[nearest_neighbors]   

    # Return the most common label among the k nearest neighbors
    predictions[i] =  np.argmax(np.bincount(nearest_neighbor_labels))


  # Return the predictions array
  return predictions

# Set the values of k
ks = [1, 2, 3, 4, 5]

# Initialize the accuracy array
accuracies = np.zeros(len(ks))

# Start the timer
start_time = time.time()

# Classify all of the test samples for each value of k
for i, k in enumerate(ks):
  for x_batch, y_batch in load_data_in_chunks(x_train, y_train, 128):
    predictions = classify_all_samples(x_batch, y_batch, k)
    accuracies[i] += np.mean(predictions == y_batch)

# Stop the timer
end_time = time.time()

# Print the classification accuracies and time required to classify all the test samples
for i, k in enumerate(ks):
  print('Classification accuracy for k = {}: {}'.format(k, accuracies[i] / len(x_test)))
  print('Time required to classify all the test samples for k = {}: {}'.format(k, end_time - start_time))


In [ ]:
x_train.shape

(33600, 785)

In [ ]:
y_train.shape

(33600,)

In [ ]:
x_test.shape

(8400, 785)

In [ ]:
y_test.shape

(8400,)

In [ ]:
import numpy as np
import time
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.feature_extraction.text import HashingVectorizer

x_train, x_test, y_train, y_test = train_test_split(X_train, Y_train, test_size=0.2, random_state=42)

# Define the LSH classifier
def lshknn(train_features, train_labels, test_features, length, s):
  # Create a hashing vectorizer with the specified length and s
  vectorizer = HashingVectorizer(n_features=length, stop_words='english', binary=True, norm='l2', ngram_range=(1, 1))

  # Create the training and test signatures
  train_signatures = vectorizer.fit_transform(train_features)
  test_signatures = vectorizer.fit_transform(test_features)
  
  # Create a kNN classifier
  clf = KNeighborsClassifier(n_neighbors=5)

  print("train_signatures - " , train_signatures.shape, "train_labels - ", train_labels.shape)

  # Train the classifier
  clf.fit_transform(train_signatures, train_labels)

  # Classify the test samples
  predictions = clf.predict(test_signatures)

  # Return the predictions
  return predictions

# Calculate the classification accuracy and time for each combination of length and s
accuracies = []
times = []
for length in [40, 60]:
  for s in [0.8, 0.9]:
    # Run the experiment multiple times
    for i in range(10):
      # Classify the test samples
      start = time.time()
      predictions = lshknn(x_train, y_train, x_test, length, s)
      end = time.time()

      # Calculate the accuracy
      accuracy = np.mean(predictions == y_test)

      # Save the accuracy and time
      accuracies.append(accuracy)
      times.append(end - start)

    # Print the classification accuracy and time
    print("length = {}, s = {}: accuracy = {:.2f}, time = {:.3f}s".format(length, s, np.mean(accuracies), np.std(accuracies)))

train_signatures -  (785, 40) train_labels -  (33600,)


ValueError: ignored

In [17]:
X_train.shape

(42000, 785)

In [16]:
Y_train.shape

(42000,)

In [44]:
y_train.to_numpy()

array([6, 5, 3, ..., 2, 6, 0])

In [16]:
import numpy as np
import time
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.feature_extraction.text import HashingVectorizer
from sklearn.feature_selection import SelectKBest, chi2

x_train, x_test, y_train, y_test = train_test_split(X_train, Y_train, test_size=0.2, random_state=42, shuffle=True)

# Define the LSH classifier
def lshknn(train_features, train_labels, test_features):
  # Create a hashing vectorizer with 40 features
  vectorizer = HashingVectorizer(stop_words='english', binary=True, norm='l2', ngram_range=(1, 1))

  # Create the training and test signatures
  train_signatures = vectorizer.fit_transform(train_features)
  test_signatures = vectorizer.transform(test_features)

  print("train_features shape:", train_features.shape)
  print("train_labels shape:", train_labels.shape)

  # Create a kNN classifier
  clf = KNeighborsClassifier(n_neighbors=5)
  print(list(train_signatures.toarray()))
  #print(list(np.array(train_signatures[0:50])))
  print("type train_labels - " , type(train_labels))
  print(list(train_labels))
  # Train the classifier
  clf.fit(list(train_signatures), list(train_labels))

  # Classify the test samples
  predictions = clf.predict(test_signatures)

  # Return the predictions
  return predictions

# Calculate the classification accuracy
accuracies = []
for i in range(10):
  # Classify the test samples
  start = time.time()
  predictions = lshknn(x_train, y_train.to_numpy(), x_test)
  end = time.time()

  # Calculate the accuracy
  accuracy = np.mean(predictions == y_test)

  # Save the accuracy
  accuracies.append(accuracy)

# Print the classification accuracy
print("Accuracy: {:.2f}".format(np.mean(accuracies)))


train_features shape: (33600, 785)
train_labels shape: (33600,)
[array([0., 0., 0., ..., 0., 0., 0.]), array([0., 0., 0., ..., 0., 0., 0.]), array([0., 0., 0., ..., 0., 0., 0.]), array([0., 0., 0., ..., 0., 0., 0.]), array([0., 0., 0., ..., 0., 0., 0.]), array([0., 0., 0., ..., 0., 0., 0.]), array([0., 0., 0., ..., 0., 0., 0.]), array([0., 0., 0., ..., 0., 0., 0.]), array([0., 0., 0., ..., 0., 0., 0.]), array([0., 0., 0., ..., 0., 0., 0.]), array([0., 0., 0., ..., 0., 0., 0.]), array([0., 0., 0., ..., 0., 0., 0.]), array([0., 0., 0., ..., 0., 0., 0.]), array([0., 0., 0., ..., 0., 0., 0.]), array([0., 0., 0., ..., 0., 0., 0.]), array([0., 0., 0., ..., 0., 0., 0.]), array([0., 0., 0., ..., 0., 0., 0.]), array([0., 0., 0., ..., 0., 0., 0.]), array([0., 0., 0., ..., 0., 0., 0.]), array([0., 0., 0., ..., 0., 0., 0.]), array([0., 0., 0., ..., 0., 0., 0.]), array([0., 0., 0., ..., 0., 0., 0.]), array([0., 0., 0., ..., 0., 0., 0.]), array([0., 0., 0., ..., 0., 0., 0.]), array([0., 0., 0., ...,

ValueError: ignored

In [18]:
from sklearn import datasets

# load the MNIST digits dataset
mnist = datasets.load_digits()

from sklearn.model_selection import train_test_split
import numpy as np
# Training and testing split,
# 75% for training and 25% for testing
(trainData, testData, trainLabels, testLabels) = train_test_split(np.array(mnist.data), mnist.target, test_size=0.25, random_state=42)

# take 10% of the training data and use that for validation
(trainData, valData, trainLabels, valLabels) = train_test_split(trainData, trainLabels, test_size=0.1, random_state=84)

In [40]:
trainData

array([[ 0.,  0.,  6., ...,  1.,  0.,  0.],
       [ 0.,  0.,  7., ..., 16., 16.,  5.],
       [ 0.,  0.,  9., ...,  6.,  0.,  0.],
       ...,
       [ 0.,  0.,  2., ...,  5.,  0.,  0.],
       [ 0.,  0.,  0., ...,  9.,  0.,  0.],
       [ 0.,  1., 12., ...,  8.,  1.,  0.]])

In [41]:
trainLabels

array([9, 1, 5, ..., 0, 1, 3])

In [26]:
print(type(trainData))

<class 'numpy.ndarray'>


In [21]:
from sklearn.neighbors import KNeighborsClassifier
kVals = range(1, 30, 2)
accuracies = []

# loop over kVals
for k in kVals:
    print("k - ", k)
    # train the classifier with the current value of `k`
    model = KNeighborsClassifier(n_neighbors=k)
    model.fit(trainData, trainLabels)

    # evaluate the model and print the accuracies list
    score = model.score(valData, valLabels)
    print("k=%d, accuracy=%.2f%%" % (k, score * 100))
    accuracies.append(score)

k -  1
k=1, accuracy=99.26%
k -  3
k=3, accuracy=99.26%
k -  5
k=5, accuracy=99.26%
k -  7
k=7, accuracy=99.26%
k -  9
k=9, accuracy=99.26%
k -  11
k=11, accuracy=99.26%
k -  13
k=13, accuracy=99.26%
k -  15
k=15, accuracy=99.26%
k -  17
k=17, accuracy=98.52%
k -  19
k=19, accuracy=98.52%
k -  21
k=21, accuracy=97.78%
k -  23
k=23, accuracy=97.04%
k -  25
k=25, accuracy=97.78%
k -  27
k=27, accuracy=97.04%
k -  29
k=29, accuracy=97.04%
